## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [14]:
import pandas as pd

df = pd.read_csv('client_dataset.csv')
df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year         item_id     category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   
6   8080280  2023-04-19          16        2023  UAC50632-25-4E  consumables   
7   8946782  2023-02-14           7        2023  UEU25316-01-1U  consumables   
8   9394118  2023-05-13          19        2023  DDA06248-98-7Z  consumables   
9   9665852  2023-04-14          15        2023  DUU85042-87-5E    furniture   

          subcategory  unit_price  unit_cost  unit_weight  qty  line_number  
0            wall art     1096.80     762.71         7.50  105            1  
1                pens       24.95      15.09         1.49   21            0  
2  project management       13.52       7.86         1.68   39            6  
3                pens       36.42      24.85         1.23   29            3  
4                misc      195.10     108.17        46.43   20            1  
5              plants      172.84     154.44         2.79   78            0  
6    kitchen supplies      245.56     231.95         1.73   56            5  
7                pens       20.99      16.97         5.12   22            4  
8      paper products       83.86      59.00         0.27   49            3  
9             seating       66.88      54.15         2.47   78            1

In [3]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [4]:
# Use the describe function to gather some basic statistics
print(df.describe())

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 19 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
 18  subtotal     54639 non-null  float64
dtypes: f

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.iloc[0:10] 
df.tail

<bound method NDFrame.tail of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  or

In [12]:
# df[['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
#        'order_date', 'order_week', 'order_year', 'item_id', 'category',
#        'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
#        'line_number']].value_counts().head(1)

first  last      job                             phone              email                      client_id  order_id  order_date  order_week  order_year  item_id         category   subcategory  unit_price  unit_cost  unit_weight  qty  line_number
Aaron  Martinez  Armed forces technical officer  283.652.2505x6088  aaron.martinez@morton.biz  19820      2178609   2023-05-01  18          2023        ABX14990-37-1Z  furniture  tables       0.02        0.01       4.22         29   1              1
Name: count, dtype: int64

In [16]:
# What three item categories had the most entries?
item_categories = df['category'].value_counts().head(3)
print(item_categories)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [38]:
# For the category with the most entries, which subcategory had the most entries?
most_common_category = item_categories.index[0]
print(most_common_category)
subcategory_counts = df[df['category'] == most_common_category]['subcategory'].value_counts()
print(subcategory_counts.head(1))

consumables
subcategory
bathroom supplies    6424
Name: count, dtype: int64


In [26]:
# Which five clients had the most entries in the data?
client_counts = df['client_id'].value_counts().head(5)
print(client_counts)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [32]:
# Store the client ids of those top 5 clients in a list.
top_clients = client_counts.index.tolist()
top_clients

[33615, 66037, 46820, 38378, 24741]

In [36]:
# How many total units (the qty column) did the client with the most entries order order?
client_with_most_entries = client_counts.index[0]
print(client_with_most_entries)

total_units = df[df['client_id'] == client_with_most_entries]['qty'].sum()
print(total_units)

33615
64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [44]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  
0                1  115164.00  
1                0     523.95  
2                6     527.28  
3                3    1056.18  
4                1    3902.00  
...            ...        ...  
54634            8    2743.29  
54635            0    9709.73  
54636            6   31188.50  
54637            0     165.76  
54638            5    3103.31  

[54639 rows x 19 columns]

In [50]:
# Find the total weight of each shipment 
# Create a column for total weight
df['total_weight'] = df['qty'] * df['unit_weight']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  total_weight  
0                1  115164.00        787.50  
1                0     523.95         31.29  
2                6     527.28         65.52  
3                3    1056.18         35.67  
4                1    3902.00        928.60  
...            ...        ...           ...  
54634            8    2743.29         74.25  
54635            0    9709.73        549.90  
54636            6   31188.50       1976.00  
54637            0     165.76       2020.48  
54638            5    3103.31       2134.17  

In [54]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping_price'] = [w * 7 if w > 50 else w * 10 for w in df['total_weight']]
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...     category         subcategory  \
0              17        2023  ...        decor            wall art   
1              20        2023  ...  consumables                pens   
2               4        2023  ...     software  project management   
3              17        2023  ...  consumables                pens   
4              21        2023  ...  consumables                misc   
...           ...         ...  ...          ...                 ...   
54634          17        2023  ...     software           logistics   
54635           6        2023  ...  consumables    kitchen supplies   
54636           9        2023  ...  consumables   bathroom supplies   
54637           3        2023  ...    furniture              stands   
54638           1        2023  ...     software  project management   

      unit_price  unit_cost  unit_weight   qty  line_number   subtotal  \
0        1096.80     762.71         7.50   105            1  115164.00   
1          24.95      15.09         1.49    21            0     523.95   
2          13.52       7.86         1.68    39            6     527.28   
3          36.42      24.85         1.23    29            3    1056.18   
4         195.10     108.17        46.43    20            1    3902.00   
...          ...        ...          ...   ...          ...        ...   
54634      83.13      51.60         2.25    33            8    2743.29   
54635     206.59     175.46        11.70    47            0    9709.73   
54636      65.66      57.31         4.16   475            6   31188.50   
54637       1.48       1.22        18.04   112            0     165.76   
54638       3.01       1.61         2.07  1031            5    3103.31   

       total_weight  shipping_price  
0            787.50         5512.50  
1             31.29          312.90  
2             65.52          458.64  
3             35.67          356.70  
4            928.60         6500.20  
...             ...             ...  
54634         74.25          519.75  
54635        549.90         3849.30  
54636       1976.00        13832.00  
54637       2020.48        

In [66]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = df['subtotal'] + df['shipping_price']
df['total_price'] =round(df['total_price'] * 1.0925, 2)
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number   subtotal  total_weight  \
0             7.50   105            1  115164.00        787.50   
1             1.49    21            0     523.95         31.29   
2             1.68    39            6     527.28         65.52   
3             1.23    29            3    1056.18         35.67   
4            46.43    20            1    3902.00        928.60   
...            ...   ...          ...        ...           ...   
54634         2.25    33            8    2743.29         74.25   
54635        11.70    47            0    9709.73        549.90   
54636         4.16   475            6   31188.50       1976.00   
54637        18.04   112            0     165.76       2020.48   
54638         2.07  1031            5    3103.31       2134.17   

       shipping_price  total_price  
0             5512.50    131839.08  
1              312.90       914.26  
2              458.64      1077.12  
3              356.70      1543.57  
4             6500.20     11364.40  
...               ...          ...  
54634          519.75      3564.87  
54635         3849.30     14813.24  
54636        13832.00     49184.90  
54637        14143.36     15632.71  


In [68]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   subtotal  total_weight  shipping_price  total_price  \
0                1  115164.00        787.50         5512.50    131839.08   
1                0     523.95         31.29          312.90       914.26   
2                6     527.28         65.52          458.64      1077.12   
3                3    1056.18         35.67          356.70      1543.57   
4                1    3902.00        928.60         6500.20     11364.40   
...            ...        ...           ...             ...          ...   
54634            8    2743.29         74.25          519.75      3564.87   
54635            0    9709.73        549.90         3849.30     14813.24   
54636            6   31188.50       1976.00        13832.00     49184.90   
54637            0     165.76       2020.48        14143.36     15632.71   
54638            5    3103.31       2134.17        14939.19     19711.43   

       line_cost  
0       85597.05  
1         629.79  
2         765.18  
3        1077.35  
4        8663.60  
...          ...  
54634    2222.55  
54635   12095.92  
54636   41054.25  
54637   14280.00  
54638   16599.10  

[54639 rows x 23 columns]

In [72]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

        subtotal  total_weight  shipping_price  total_price  line_cost  \
0      115164.00        787.50         5512.50    131839.08   85597.05   
1         523.95         31.29          312.90       914.26     629.79   
2         527.28         65.52          458.64      1077.12     765.18   
3        1056.18         35.67          356.70      1543.57    1077.35   
4        3902.00        928.60         6500.20     11364.40    8663.60   
...          ...           ...             ...          ...        ...   
54634    2743.29         74.25          519.75      3564.87    2222.55   
54635    9709.73        549.90         3849.30     14813.24   12095.92   
54636   31188.50       1976.00        13832.00     49184.90   41054.25   
54637     165.76       2020.48        14143.36     15632.71   14280.00   
54638    3103.31       2134.17        14939.19     19711.43   16599.10   

       line_profit  
0         46242.03  
1           284.47  
2           311.94  
3           466.22  
4          2700.80  
...            ...  
54634      1342.32  
54635      2717.32  
54636      8130.65  
54637      1352.71  
54638      3112.33  

[54639 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [74]:
# Check your work using the totals above
receipts = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}
for order_id, receipt_total in receipts.items():
    order_df = df[df['order_id'] == order_id]
    order_total = order_df['total_price'].sum()
    print(f"Order {order_id}: {order_total:.2f} (expected {receipt_total:.2f})")

Order 2742071: 152811.89 (expected 152811.89)
Order 2173913: 162388.71 (expected 162388.71)
Order 6128929: 923441.25 (expected 923441.25)


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [94]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids. 
top_client_spending = {}
for client_id in top_clients:
    client_df = df[df['client_id'] == client_id]
    top_client_spending[client_id] = {
        'total_units': client_df['qty'].sum(),
        'total_shipping_price': client_df['shipping_price'].sum(),
        'total_revenue': client_df['total_price'].sum(),
        'total_profit': client_df['line_profit'].sum()
    }
print(top_client_spending)

{33615: {'total_units': 64313, 'total_shipping_price': 1828984.89, 'total_revenue': 8377308.520000001, 'total_profit': 2201994.61}, 66037: {'total_units': 43018, 'total_shipping_price': 1395151.8499999999, 'total_revenue': 10259514.79, 'total_profit': 3255031.8100000005}, 46820: {'total_units': 75768, 'total_shipping_price': 1601448.8399999999, 'total_revenue': 9743794.36, 'total_profit': 2736602.7199999997}, 38378: {'total_units': 73667, 'total_shipping_price': 3429455.4000000004, 'total_revenue': 12906550.869999997, 'total_profit': 3271829.8899999997}, 24741: {'total_units': 239862, 'total_shipping_price': 5126448.37, 'total_revenue': 82268892.02000001, 'total_profit': 36579992.309999995}}


In [100]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = pd.DataFrame.from_records(top_client_spending).T.reset_index()
summary_df.columns = ['Client ID', 'Total Units', 'Total Shipping Price', 'Total Revenue', 'Total Profit']
summary_df

Client ID  Total Units  Total Shipping Price  Total Revenue  Total Profit
0      24741     239862.0            5126448.37    82268892.02   36579992.31
1      33615      64313.0            1828984.89     8377308.52    2201994.61
2      38378      73667.0            3429455.40    12906550.87    3271829.89
3      46820      75768.0            1601448.84     9743794.36    2736602.72
4      66037      43018.0            1395151.85    10259514.79    3255031.81

In [130]:
# Format the data and rename the columns to names suitable for presentation.
# Define the money columns. 
money_columns = ['Total Revenue', 'Total Profit']

# Define a function that converts a dollar amount to millions.
def format_currency(x):
    # Convert the string to a float after removing dollar signs and commas
    # numeric_value = float(x.replace('$', '').replace(',', ''))
    return "${:.2f}M".format(numeric_value / 1e6)

# Apply the currency_format_millions function to only the money columns. 
for column in money_columns:
    if column in summary_df.columns:
        summary_df[column] = summary_df[column].apply(format_currency)

# Rename the columns to reflect the change in the money format. 
summary_df.rename(columns={
    'Total Revenue': 'Total Revenue (Millions)',
    'Total Profit': 'Total Profit (Millions)',
}, inplace=True)

# Print the formatted DataFrame to verify changes
print(summary_df)


NameError: name 'numeric_value' is not defined

In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
summary_df.sort_values('Total Profit', ascending=False, inplace=True)
print(summary_df)

In [ ]:
summary = f"The top 5 clients by quantity spent a total of {summary_df['Total Revenue'].sum():.2f} million dollars. The top client by profit is {summary_df.iloc[0]['Client ID']} with a profit of {summary_df.iloc[0]['Total Profit']:,.2f} million dollars."
print(sorted_summary_df)


In [ ]:
# a brief 2-3 sentence summary of my findings:
